### PR TITLE
Change the contract of get_typesupport_handle_function.

### PR DESCRIPTION
**Pull request by clalancette**
Before it would set an RCUTILS error if the typesupport
didn't match.  But because this is essentially a probe, we
change it to not set that on error and instead just return
a nullptr.  Callers can disambiguate between the probe not
finding a typesupport and a real error by checking the return
value for nullptr and then checking if `rcutils_error_is_set`.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

I went ahead and checked all of the callers of these functions that I could find, and as far as I can tell all of them will be OK with this change.  There may be some additional tests that are expecting the old error; besides the ones I fixed here, I didn't find any additional ones.  But I didn't run them all, so the full CI run should help track that down.

This change should help resolve some of the errors on CentOS that occurred in CI in combination with https://github.com/ros2/rmw/pull/293 .  This is still a draft PR since I'm not 100% sure we want to do this; opinions welcome.

### CODE REVIEW

**Comment by clalancette** [comment]
Full CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13356/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8283/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11085/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13407/)
* CentOS [Build Status](http://ci.ros2.org/job/ci_linux-centos/96/)

File: rosidl_typesupport_c/src/type_support_dispatch.hpp [hunk 1]
```diff
@@ -34,85 +33,103 @@ namespace rosidl_typesupport_c
...
-          "Failed to get symbol '%s' in library: %s",
-          map->symbol_name[i], e.what());
+          "Could not load library %s: %s", library_name.c_str(), e.what());
+        return nullptr;
+      } catch (const std::bad_alloc & e) {
+        RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+          "Could not load library %s: %s", library_name.c_str(), e.what());
         return nullptr;
       }
+      map->data[i] = lib;
```
Comment (sloretz) [inline rosidl_typesupport_c/src/type_support_dispatch.hpp:106]: Any explanation of how this code works that you're aware of? It's a little strange that it modifies a non-const `void **` member of a `const type_support_t *`.

**Review by sloretz — COMMENTED** [review]
The change to not set an error seems fine, though it's hard to judge because the purpose of the existing code is not clear to me.

The unrelated change to the first conditional change seems fine, but it makes the diff pretty hard to follow.

File: rosidl_typesupport_c/src/type_support_dispatch.hpp [hunk 2]
```diff
@@ -34,85 +33,103 @@ namespace rosidl_typesupport_c
...
+        return nullptr;
       }
-      auto clib = static_cast<const rcpputils::SharedLibrary *>(map->data[i]);
-      lib = const_cast<rcpputils::SharedLibrary *>(clib);
 
-      void * sym = nullptr;
+      std::string library_name;
+      try {
+        library_name = rcpputils::get_platform_library_name(library_basename);
+      } catch (const std::runtime_error & e) {
```
[the remark this answers is not in the document]
Comment (clalancette) [reply rosidl_typesupport_c/src/type_support_dispatch.hpp:88]: Yes, exactly.

File: rosidl_typesupport_c/src/type_support_dispatch.hpp (hunk 1 above)
Comment (clalancette) [reply rosidl_typesupport_c/src/type_support_dispatch.hpp:106]: Honestly, I just moved it around.  @hidmic was involved in the latest rewrite of it, I believe.  Maybe he can shed some light.
